presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java

```diff
@@ -0,0 +1,246 @@
+import com.facebook.presto.sql.tree.LambdaExpression;
+import com.facebook.presto.sql.tree.Literal;
+import com.facebook.presto.sql.tree.LongLiteral;
+import com.facebook.presto.sql.tree.NodeLocation;
+import com.facebook.presto.sql.tree.NotExpression;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.SearchedCaseExpression;
+import com.facebook.presto.sql.tree.WhenClause;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Optional;
+
+public class SimplifyArrayOperationsRewriter
+{
+    private SimplifyArrayOperationsRewriter() {}
+
+    public static Expression simplifyArrayOperations(Expression expression, Session session)
+    {
+        if (SystemSessionProperties.isSimplifyArrayOperations(session)) {
+            return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression);
+        }
+
+        return expression;
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private static boolean isConstant(Expression expression)
+        {
+            if (expression instanceof Cast) {
+                return isConstant(((Cast) expression).getExpression());
+            }
+
+            return expression instanceof Literal;
+        }
+
+        private static boolean isZero(Expression expression)
+        {
+            if (expression instanceof Cast) {
+                return isZero(((Cast) expression).getExpression());
+            }
+
+            return expression instanceof LongLiteral && ((LongLiteral) expression).getValue() == 0;
+        }
+
+        private static boolean isCardinalityOfFilter(Expression expression)
+        {
+            if (expression instanceof FunctionCall) {
+                FunctionCall functionCall = (FunctionCall) expression;
+                return functionCall.getName().toString().equals("cardinality") &&
+                        functionCall.getArguments().size() == 1 &&
+                        functionCall.getArguments().get(0) instanceof FunctionCall &&
+                        ((FunctionCall) (functionCall.getArguments().get(0))).getName().toString().equals("filter");
+            }
+
+            return false;
+        }
+
+        private static boolean isSimpleComaparison(ComparisonExpression comparisonExpression)
+        {
+            switch (comparisonExpression.getOperator()) {
+                case EQUAL:
+                case GREATER_THAN:
+                case LESS_THAN:
+                case NOT_EQUAL:
+                case GREATER_THAN_OR_EQUAL:
+                case LESS_THAN_OR_EQUAL:
+                    return true;
+            }
+
+            return false;
+        }
+
+        private Expression simplifyCardinalityOfFilterComparedToZero(
+                NodeLocation location,
+                Expression arrayArg,
+                ComparisonExpression.Operator operator,
+                LambdaArgumentDeclaration origLambdaArgument,
```

Comment:
In Presto most of the time we use full name, so let's call this `originalLambdaArgument`